File: LeetCode/Arrays/2091-Removing-Minimum-and-Maximum-From-Array.py
"""
Problem: 2091. Removing Minimum and Maximum From Array
Platform: LeetCode
Difficulty: Medium

Approach:
1. Find the indices of the minimum and maximum elements.
2. Let left = smaller index and right = larger index.
3. There are only three possibilities:
   - Remove both from the front.
   - Remove both from the back.
   - Remove one from each side.
4. Return the minimum of the three.

Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import List


class Solution:
    def minimumDeletions(self, nums: List[int]) -> int:
        n = len(nums)

        min_idx = 0
        max_idx = 0

        for i in range(n):
            if nums[i] < nums[min_idx]:
                min_idx = i

            if nums[i] > nums[max_idx]:
                max_idx = i

        left = min(min_idx, max_idx)
        right = max(min_idx, max_idx)

        # Both from the front
        from_front = right + 1

        # Both from the back
        from_back = n - left

        # One from front and one from back
        from_both = left + 1 + n - right

        return min(from_front, from_back, from_both)

File: LeetCode/Arrays/2091-Removing-Minimum-and-Maximum-From-Array.js
/**
 * Problem: 2091. Removing Minimum and Maximum From Array
 * Platform: LeetCode
 * Difficulty: Medium
 *
 * Approach:
 * 1. Find the indices of the minimum and maximum elements.
 * 2. Let left = smaller index and right = larger index.
 * 3. Consider three possibilities:
 *    - Both from the front.
 *    - Both from the back.
 *    - One from each side.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * @param {number[]} nums
 * @return {number}
 */
var minimumDeletions = function(nums) {
    const n = nums.length;

    let minIdx = 0;
    let maxIdx = 0;

    for (let i = 0; i < n; i++) {
        if (nums[i] < nums[minIdx]) {
            minIdx = i;
        }

        if (nums[i] > nums[maxIdx]) {
            maxIdx = i;
        }
    }

    const left = Math.min(minIdx, maxIdx);
    const right = Math.max(minIdx, maxIdx);

    // Both from the front
    const fromFront = right + 1;

    // Both from the back
    const fromBack = n - left;

    // One from front and one from back
    const fromBoth = left + 1 + n - right;

    return Math.min(fromFront, fromBack, fromBoth);
};